File: M3C-High-Performance-Computing/hw3/hw3.py
"""M3C 2018 Homework 3 01203261 Jimmy Yeung
Contains five functions:
    plot_S: plots S matrix -- use if you like
    simulate2: Simulate tribal competition over m trials. Return: all s matrices at final time
        and fc at nt+1 times averaged across the m trials.
    performance: To be completed -- analyze and assess performance of python, fortran, and fortran+openmp simulation codes
    analyze: To be completed -- analyze influence of model parameter, g
    visualize: To be completed -- generate animation illustrating "interesting" tribal dynamics
"""
import numpy as np
import matplotlib.pyplot as plt
from m1 import tribes as tr #assumes that hw3_dev.f90 has been compiled with: f2py3 --f90flags='-fopenmp' -c hw3_dev.f90 -m m1 -lgomp
#May also use scipy and time modules as needed
import time
import matplotlib.animation as animation


def plot_S(S):
    """Simple function to create plot from input S matrix
    """
    ind_s0 = np.where(S==0) #C locations
    ind_s1 = np.where(S==1) #M locations
    plt.plot(ind_s0[1],ind_s0[0],'rs')
    plt.plot(ind_s1[1],ind_s1[0],'bs')
    plt.show()
    return None
#------------------

def plot(S):
    """Similar to plot_S. Created so that I can use it in my updatefig function in visualise.
    """
    ind_s0 = np.where(S==0) #C locations
    ind_s1 = np.where(S==1) #M locations
    plot = plt.plot(ind_s0[1],ind_s0[0],'rs', ind_s1[1],ind_s1[0],'bs')
    return plot

def simulate2(N,Nt,b,e,g,m):
    """Simulate m trials of C vs. M competition on N x N grid over
    Nt generations. b, e, and g are model parameters
    to be used in fitness calculations.
    Output: S: Status of each gridpoint at end of simulation, 0=M, 1=C
            fc_ave: fraction of villages which are C at all Nt+1 times
                    averaged over the m trials
    """
    #Set initial condition
    S  = np.ones((N,N,m),dtype=int) #Status of each gridpoint: 0=M, 1=C
    j = int((N-1)/2)
    S[j,j,:] = 0
    N2inv = 1./(N*N)

    fc_ave = np.zeros(Nt+1) #Fraction of points which are C
    fc_ave[0] = S.sum()

    #Initialize matrices
    NB = np.zeros((N,N,m),dtype=int) #Number of neighbors for each point
    NC = np.zeros((N,N,m),dtype=int) #Number of neighbors who are Cs
    S2 = np.zeros((N+2,N+2,m),dtype=int) #S + border of zeros
    F = np.zeros((N,N,m)) #Fitness matrix
    F2 = np.zeros((N+2,N+2,m)) #Fitness matrix + border of zeros
    A = np.ones((N,N,m)) #Fitness parameters, each of N^2 elements is 1 or b
    P = np.zeros((N,N,m)) #Probability matrix
    Pden = np.zeros((N,N,m))
    #---------------------

    #Calculate number of neighbors for each point
    NB[:,:,:] = 8
    NB[0,1:-1,:],NB[-1,1:-1,:],NB[1:-1,0,:],NB[1:-1,-1,:] = 5,5,5,5
    NB[0,0,:],NB[-1,-1,:],NB[0,-1,:],NB[-1,0,:] = 3,3,3,3
    NBinv = 1.0/NB
    #-------------

    #----Time marching-----
    for t in range(Nt):
        R = np.random.rand(N,N,m) #Random numbers used to update S every time step

        #Set up coefficients for fitness calculation
        A = np.ones((N,N,m))
        ind0 = np.where(S==0)
        A[ind0] = b

        #Add boundary of zeros to S
        S2[1:-1,1:-1,:] = S

        #Count number of C neighbors for each point
        NC = S2[:-2,:-2,:]+S2[:-2,1:-1,:]+S2[:-2,2:,:]+S2[1:-1,:-2,:] + S2[1:-1,2:,:] + S2[2:,:-2,:] + S2[2:,1:-1,:] + S2[2:,2:,:]

        #Calculate fitness matrix, F----
        F = NC*A
        F[ind0] = F[ind0] + (NB[ind0]-NC[ind0])*e
        F = F*NBinv
        #-----------

        #Calculate probability matrix, P-----
        F2[1:-1,1:-1,:]=F
        F2S2 = F2*S2
        #Total fitness of cooperators in community
        P = F2S2[:-2,:-2,:]+F2S2[:-2,1:-1,:]+F2S2[:-2,2:,:]+F2S2[1:-1,:-2,:] + F2S2[1:-1,1:-1,:] + F2S2[1:-1,2:,:] + F2S2[2:,:-2,:] + F2S2[2:,1:-1,:] + F2S2[2:,2:,:]

        #Total fitness of all members of community
        Pden = F2[:-2,:-2,:]+F2[:-2,1:-1,:]+F2[:-2,2:,:]+F2[1:-1,:-2,:] + F2[1:-1,1:-1,:] + F2[1:-1,2:,:] + F2[2:,:-2,:] + F2[2:,1:-1,:] + F2[2:,2:,:]

        P = (P/Pden)*g + 0.5*(1.0-g) #probability matrix
        #---------

        #Set new affiliations based on probability matrix and random numbers stored in R
        S[:,:,:] = 0
        S[R<=P] = 1

        fc_ave[t+1] = S.sum()
        #----Finish time marching-----

    fc_ave = fc_ave*N2inv/m

    return S,fc_ave
#------------------

def simulate3(N,Nt,b,e,g,m):
    "Simulate3 has been created to return an array of the fc at every year. This is used in visualize.
    "

    #Set initial condition
    S  = np.ones((N,N,m),dtype=int) #Status of each gridpoint: 0=M, 1=C
    j = int((N-1)/2)
    S[j,j,:] = 0
    N2inv = 1./(N*N)

    #Initialize matrices
    NB = np.zeros((N,N,m),dtype=int) #Number of neighbors for each point
    NC = np.zeros((N,N,m),dtype=int) #Number of neighbors who are Cs
    S2 = np.zeros((N+2,N+2,m),dtype=int) #S + border of zeros
    F = np.zeros((N,N,m)) #Fitness matrix
    F2 = np.zeros((N+2,N+2,m)) #Fitness matrix + border of zeros
    A = np.ones((N,N,m)) #Fitness parameters, each of N^2 elements is 1 or b
    P = np.zeros((N,N,m)) #Probability matrix
    Pden = np.zeros((N,N,m))
    #---------------------

    #Calculate number of neighbors for each point
    NB[:,:,:] = 8
    NB[0,1:-1,:],NB[-1,1:-1,:],NB[1:-1,0,:],NB[1:-1,-1,:] = 5,5,5,5
    NB[0,0,:],NB[-1,-1,:],NB[0,-1,:],NB[-1,0,:] = 3,3,3,3
    NBinv = 1.0/NB
    #-------------
    S_evo = np.zeros((N,N,Nt+1))
    S_evo[:,:,0] = S[:,:,0]

    #----Time marching-----
    for t in range(Nt):
        R = np.random.rand(N,N,m) #Random numbers used to update S every time step

        #Set up coefficients for fitness calculation
        A = np.ones((N,N,m))
        ind0 = np.where(S==0)
        A[ind0] = b

        #Add boundary of zeros to S
        S2[1:-1,1:-1,:] = S

        #Count number of C neighbors for each point
        NC = S2[:-2,:-2,:]+S2[:-2,1:-1,:]+S2[:-2,2:,:]+S2[1:-1,:-2,:] + S2[1:-1,2:,:] + S2[2:,:-2,:] + S2[2:,1:-1,:] + S2[2:,2:,:]

        #Calculate fitness matrix, F----
        F = NC*A
        F[ind0] = F[ind0] + (NB[ind0]-NC[ind0])*e
        F = F*NBinv
        #-----------

        #Calculate probability matrix, P-----
        F2[1:-1,1:-1,:]=F
        F2S2 = F2*S2
        #Total fitness of cooperators in community
        P = F2S2[:-2,:-2,:]+F2S2[:-2,1:-1,:]+F2S2[:-2,2:,:]+F2S2[1:-1,:-2,:] + F2S2[1:-1,1:-1,:] + F2S2[1:-1,2:,:] + F2S2[2:,:-2,:] + F2S2[2:,1:-1,:] + F2S2[2:,2:,:]

        #Total fitness of all members of community
        Pden = F2[:-2,:-2,:]+F2[:-2,1:-1,:]+F2[:-2,2:,:]+F2[1:-1,:-2,:] + F2[1:-1,1:-1,:] + F2[1:-1,2:,:] + F2[2:,:-2,:] + F2[2:,1:-1,:] + F2[2:,2:,:]

        P = (P/Pden)*g + 0.5*(1.0-g) #probability matrix
        #---------

        #Set new affiliations based on probability matrix and random numbers stored in R
        S[:,:,:] = 0
        S[R<=P] = 1

        S_evo[:,:,t+1] = S[:,:,0]

    return S_evo

def performance(input=(None),display=False):
    """I wanted to analyse the speedup of Fortran+OMP against Fortran as the no. of years, Nt, increases for a fixed m = 100.
    Looking at hw321, we can see that the speedup is between 1.6 and 1.85 for all years 10 to 200.
    This agrees with the fact that the speedup will be less than the number of threads used, which in this case is 2.

    In hw322, I have fixed m = 100 and plotted the time it takes for the functions to run against
    the number of years, Nt, it is iterated over.
    I have done this to see the relative performances of Python, Fortran and Fortran+OMP.
    Python is by far the slowest - takes multiple times longer than Fortran and Fortran+OMP.
    The time it takes for Fortran is just under twice as long as Fortran+OMP - just as expected from
    our calculated speedup.
    The time each function takes is proportional to the number of years iterated.
    """

    tr.tr_b = 1.1
    tr.tr_e = 0.01
    tr.tr_g = 0.95
    tr.numthreads = 2

    #Speedup of parallelised code
    speedups = list()
    x_axis = list(range(10,210,10))
    for i in x_axis:
        t0 = time.time()
        for i in range(20):
            tr.simulate2_f90(21,i,100)
        t1 = time.time()
        time_vec = t1-t0

        t0 = time.time()
        for i in range(20):
            tr.simulate2_omp(21,i,100)
        t1 = time.time()
        time_parallel = (t1-t0)

        speedup = time_vec/time_parallel
        speedups.append(speedup)
    plt.figure()
    plt.plot(x_axis,speedups)
    plt.title("Speedup of Fortran+OMP against Fortran - m = 100")
    plt.xlabel('Nt, years')
    plt.ylabel('Time, s')
    plt.show()


    #Relative performance of Python, Fortran and Fortran+OMP
    times_for, times_omp, times_py = list(), list(), list()
    x_axis = list(range(10,110,10))
    for nt in x_axis:
         t0 = time.time()
         for i in range(20):
            tr.simulate2_f90(21,nt,100)
         t1 = time.time()
         time_for = (t1-t0)/20
         times_for.append(time_for)

         t0 = time.time()
         for i in range(20):
            tr.simulate2_omp(21,nt,100)
         t1 = time.time()
         time_omp = (t1-t0)/20
         times_omp.append(time_omp)

         t0 = time.time()
         for i in range(20):
            simulate2(21,nt,1.1,0.01,0.95,100)
         t1 = time.time()
         time_py = (t1-t0)/20
         times_py.append(time_py)
    plt.figure()
    plt.plot(x_axis,times_for, label="Fortran")
    plt.plot(x_axis,times_omp, label="Fortran + OMP")
    plt.plot(x_axis,times_py, label="Python")
    plt.title("Performance - m = 100")
    plt.xlabel('Nt, years')
    plt.ylabel('Time, s')
    plt.legend()
    plt.show()

    return None

#---------------------------
def analyze():
    """To investigate the influence off g, I plotted three graphs of different values of g for a fixed value of b.
    I plotted the % of villages which are C against the no. of years.
    We know that a larger value of b means that C decreases at a quicker rate.

    For b = 1.01, 1.25 and 1.5, the effect of g was very similar.
    For g between 0.8 and 0.95, a larger g meant that C decreases at a slower rate,
    but the value that it converges at also becomes smaller.

    For g = 1, the pattern was completely different to the other values of g.
    g = 1 meant that the rate at which C decreased was very slow, and the value of which
    C converges depends on b. As b increases, then the value at which C
    converges decreases.
    """

    x=range(200+1)

    plt.figure(figsize=(12,4))
    for g in np.linspace(0.8,1,5):
        _, fc_ave = simulate2(51,200,1.01,0.01,g,50)
        plt.plot(x, fc_ave*100, label=g)
    plt.title('Evolution of C for different values of g, b = 1.01')
    plt.xlabel('Number of years past')
    plt.ylabel('% of villages which are C')
    plt.legend()
    plt.show()

    plt.figure(figsize=(12,4))
    for g in np.linspace(0.8,1,5):
        _, fc_ave = simulate2(51,200,1.25,0.01,g,50)
        plt.plot(x, fc_ave*100, label=g)
    plt.title('Evolution of C for different values of g, b = 1.25')
    plt.xlabel('Number of years past')
    plt.ylabel('% of villages which are C')
    plt.legend()
    plt.show()

    plt.figure(figsize=(12,4))
    for g in np.linspace(0.8,1,5):
        _, fc_ave = simulate2(51,200,1.5,0.01,g,50)
        plt.plot(x, fc_ave*100, label=g)
    plt.title('Evolution of C for different values of g, b = 1.5')
    plt.xlabel('Number of years past')
    plt.ylabel('% of villages which are C')
    plt.legend()
    plt.show()

    return None #Modify as needed

#------------------------------------


def visualize():
    """Generate an animation illustrating the evolution of
        villages during C vs M competition
    """
    fig = plt.figure()
    S = simulate3(21,100,1.1,0.01,0.95,1)

    def updatefig(i):
        plt = plot(S[:,:,i])
        return plt

    ani = animation.FuncAnimation(fig, updatefig,frames=100,repeat=False,blit=True)
    ani.save("hw3movie.mp4")

    return

#------------------------------------
if __name__ == '__main__':
    #Modify the code here so that it calls performance analyze and
    #generates the figures that you are submitting with your code

    output_a = performance()

    output_b = analyze()

    output_c = visualize()

File: M3C-High-Performance-Computing/hw3/hw3.f90
!M3C 2018 Homework 3 01203261 Jimmy Yeung
!This module contains four module variables and two subroutines;
!one of these routines must be developed for this assignment.
!Module variables--
! tr_b, tr_e, tr_g: the parameters b, e, and g in the tribe competition model
! numthreads: The number of threads that should be used in parallel regions within simulate2_omp
!
!Module routines---
! simulate2_f90: Simulate tribal competition over m trials. Return: all s matrices at final time
! and fc at nt+1 times averaged across the m trials.
! simulate2_omp: Same input/output functionality as simulate2.f90 but parallelized with OpenMP

module tribes
  use omp_lib
  implicit none
  integer :: numthreads
  real(kind=8) :: tr_b,tr_e,tr_g
contains

!Simulate m trials of Cooperator vs. Mercenary competition using the parameters, tr_b and tr_e.
!Input:
! n: defines n x n grid of villages
! nt: number of time steps
! m: number of trials
!Output:
! s: status matrix at final time step for all m trials
! fc_ave: fraction of cooperators at each time step (including initial condition)
! averaged across the m trials
subroutine simulate2_f90(n,nt,m,s,fc_ave)
  implicit none
  integer, intent(in) :: n,nt,m
  integer, intent(out), dimension(n,n,m) :: s
  real(kind=8), intent(out), dimension(nt+1) :: fc_ave
  integer :: i1,j1
  real(kind=8) :: n2inv
  integer, dimension(n,n,m) :: nb,nc
  integer, dimension(n+2,n+2,m) :: s2
  real(kind=8), dimension(n,n,m) :: f,p,a,pden,nbinv
  real(kind=8), dimension(n+2,n+2,m) :: f2,f2s2
  real(kind=8), allocatable, dimension(:,:,:) :: r !random numbers

  !---Problem setup----
  !Initialize arrays and problem parameters

  !initial condition
  s=1
  j1 = (n+1)/2
  s(j1,j1,:) = 0

  n2inv = 1.d0/dble(n*n)
  fc_ave(1) = sum(s)*(n2inv/m)

  s2 = 0
  f2 = 0.d0
  !Calculate number of neighbors for each point
  nb = 8
  nb(1,2:n-1,:) = 5
  nb(n,2:n-1,:) = 5
  nb(2:n-1,1,:) = 5
  nb(2:n-1,n,:) = 5
  nb(1,1,:) = 3
  nb(1,n,:) = 3
  nb(n,1,:) = 3
  nb(n,n,:) = 3

  nbinv = 1.d0/nb
  allocate(r(n,n,m))
  !---finished Problem setup---

  !----Time marching----
  do i1=1,nt

    call random_number(r) !Random numbers used to update s every time step

    !Set up coefficients for fitness calculation in matrix, a
    a = 1
    where(s==0)
      a=tr_b
    end where

    !create s2 by adding boundary of zeros to s
    s2(2:n+1,2:n+1,:) = s

    !Count number of C neighbors for each point
    nc = s2(1:n,1:n,:) + s2(1:n,2:n+1,:) + s2(1:n,3:n+2,:) + &
         s2(2:n+1,1:n,:)                  + s2(2:n+1,3:n+2,:) + &
         s2(3:n+2,1:n,:)   + s2(3:n+2,2:n+1,:)   + s2(3:n+2,3:n+2,:)

    !Calculate fitness matrix, f----
    f = nc*a
    where(s==0)
      f = f + (nb-nc)*tr_e
    end where
    f = f*nbinv

    !Calculate probability matrix, p----
    f2(2:n+1,2:n+1,:) = f
    f2s2 = f2*s2

    !Total fitness of cooperators in community
    p = f2s2(1:n,1:n,:) + f2s2(1:n,2:n+1,:) + f2s2(1:n,3:n+2,:) + &
           f2s2(2:n+1,1:n,:) + f2s2(2:n+1,2:n+1,:)  + f2s2(2:n+1,3:n+2,:) + &
          f2s2(3:n+2,1:n,:)   + f2s2(3:n+2,2:n+1,:)   + f2s2(3:n+2,3:n+2,:)

    !Total fitness of all members of community
    pden = f2(1:n,1:n,:) + f2(1:n,2:n+1,:) + f2(1:n,3:n+2,:) + &
           f2(2:n+1,1:n,:) + f2(2:n+1,2:n+1,:)  + f2(2:n+1,3:n+2,:) + &
          f2(3:n+2,1:n,:)   + f2(3:n+2,2:n+1,:)   + f2(3:n+2,3:n+2,:)

    p = (p/pden)*tr_g + 0.5d0*(1.d0-tr_g) !probability matrix

    !Set new affiliations based on probability matrix and random numbers stored in R
    s = 0
    where (R<=p)
        s = 1
    end where

    fc_ave(i1+1) = sum(s)*(n2inv/m)
  end do
end subroutine simulate2_f90


!Simulate m trials of Cooperator vs. Mercenary competition using the parameters, tr_b and tr_e.
!Same functionality as simulate2_f90, but parallelized with OpenMP
!Parallel regions should use numthreads threads.
!Input:
! n: defines n x n grid of villages
! nt: number of time steps
! m: number of trials
!Output:
! s: status matrix at final time step for all m trials
! fc_ave: fraction of cooperators at each time step (including initial condition)
! averaged across the m trials
subroutine simulate2_omp(n,nt,m,s,fc_ave)
  !My approach to parallisation was to split the m trials over a number of threads.
  !First, I changed the vectorised code so that it calculates the m trials using a do loop.
  !Then, I used OMP parallel do to split up the do loop iterations over a number of threads - set using omp_set_num_threads.
  !It was not possible to parallelise the loop from 1 to nt because the S calculated in each year depends on the previous year.
  !So I parallelised the loop from 1 to m which is inside the loop from 1 to nt.
  implicit none
  integer, intent(in) :: n,nt,m
  integer, intent(out), dimension(n,n,m) :: s
  real(kind=8), intent(out), dimension(nt+1) :: fc_ave
  integer :: i1,j1
  real(kind=8), allocatable, dimension(:,:,:) :: r !random numbers

  !Add further variables as needed
  real(kind=8) :: n2inv
  integer, dimension(n,n,m) :: nb,nc
  integer, dimension(n+2,n+2,m) :: s2
  real(kind=8), dimension(n,n,m) :: f,p,a,pden,nbinv
  real(kind=8), dimension(n+2,n+2,m) :: f2,f2s2
  integer:: i2,j2

  !initial condition and r allocation (does not need to be parallelized)
  s=1
  j1 = (n+1)/2
  s(j1,j1,:) = 0
  allocate(r(n,n,m))
  !------------------
  n2inv = 1.d0/dble(n*n)
  fc_ave(1) = sum(s)*(n2inv/m)

  !$ call omp_set_num_threads(numthreads)

  !$OMP parallel do
  do i1 = 1, m
    s2(:,:,i1) = 0
    f2(:,:,i1) = 0.d0
    nb(:,:,i1) = 8
    nb(1,2:n-1,i1) = 5
    nb(n,2:n-1,i1) = 5
    nb(2:n-1,1,i1) = 5
    nb(2:n-1,n,i1) = 5
    nb(1,1,i1) = 3
    nb(1,n,i1) = 3
    nb(n,1,i1) = 3
    nb(n,n,i1) = 3
    nbinv(:,:,i1) = 1.d0/nb(:,:,i1)
  end do
  !$OMP end parallel do

  do i2=1,nt

    !$OMP parallel do

    do j2 = 1,m

      call random_number(r(:,:,j2))!Random numbers used to update s every time step

      a(:,:,j2) = 1
      where(s(:,:,j2)==0)
        a(:,:,j2)=tr_b
      end where

      !create s2 by adding boundary of zeros to s
      s2(2:n+1,2:n+1,j2) = s(:,:,j2)

      !Count number of C neighbors for each point
      nc(:,:,j2) = s2(1:n,1:n,j2) + s2(1:n,2:n+1,j2) + s2(1:n,3:n+2,j2) + &
                s2(2:n+1,1:n,j2)                  + s2(2:n+1,3:n+2,j2) + &
                s2(3:n+2,1:n,j2)   + s2(3:n+2,2:n+1,j2)   + s2(3:n+2,3:n+2,j2)

      f(:,:,j2) = nc(:,:,j2)*a(:,:,j2)

      where(s(:,:,j2)==0)
        f(:,:,j2) = f(:,:,j2) + (nb(:,:,j2)-nc(:,:,j2))*tr_e
      end where
      f(:,:,j2) = f(:,:,j2)*nbinv(:,:,j2)

      !Calculate probability matrix, p----
      f2(2:n+1,2:n+1,j2) = f(:,:,j2)
      f2s2(:,:,j2) = f2(:,:,j2)*s2(:,:,j2)

      !Total fitness of cooperators in community
      p(:,:,j2) = f2s2(1:n,1:n,j2) + f2s2(1:n,2:n+1,j2) + f2s2(1:n,3:n+2,j2) + &
                  f2s2(2:n+1,1:n,j2) + f2s2(2:n+1,2:n+1,j2)  + f2s2(2:n+1,3:n+2,j2) + &
                  f2s2(3:n+2,1:n,j2)   + f2s2(3:n+2,2:n+1,j2)   + f2s2(3:n+2,3:n+2,j2)

      !Total fitness of all members of community
      pden(:,:,j2) = f2(1:n,1:n,j2) + f2(1:n,2:n+1,j2) + f2(1:n,3:n+2,j2) + &
                     f2(2:n+1,1:n,j2) + f2(2:n+1,2:n+1,j2)  + f2(2:n+1,3:n+2,j2) + &
                     f2(3:n+2,1:n,j2)   + f2(3:n+2,2:n+1,j2)   + f2(3:n+2,3:n+2,j2)

      p(:,:,j2) = (p(:,:,j2)/pden(:,:,j2))*tr_g + 0.5d0*(1.d0-tr_g) !probability matrix

      s(:,:,j2) = 0
      where (R(:,:,j2)<=p(:,:,j2))
         s(:,:,j2) = 1
      end where

    end do

    !$OMP end parallel do
    fc_ave(i2+1) = sum(s)*(n2inv/m)

  end do

  deallocate(r)
end subroutine simulate2_omp

end module tribes
